clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -451,68 +486,71 @@ public int compare(ProducerBatch o1, ProducerBatch o2) {
+        if (first == null)
             return RecordBatch.NO_SEQUENCE;
 
-        return firstInFlightBatch.baseSequence();
+        return first.baseSequence();
     }
 
     synchronized ProducerBatch nextBatchBySequence(TopicPartition topicPartition) {
-        PriorityQueue<ProducerBatch> queue = inflightBatchesBySequence.get(topicPartition);
-        if (queue == null)
-            return null;
+        PriorityQueue<ProducerBatch> queue = topicPartitionBookkeeper.get(topicPartition).inflightBatchesBySequence;
         return queue.peek();
     }
 
     synchronized void removeInFlightBatch(ProducerBatch batch) {
-        PriorityQueue<ProducerBatch> queue = inflightBatchesBySequence.get(batch.topicPartition);
-        if (queue == null)
-            return;
-        queue.remove(batch);
+        if (hasInflightBatches(batch.topicPartition)) {
+            PriorityQueue<ProducerBatch> queue = topicPartitionBookkeeper.get(batch.topicPartition).inflightBatchesBySequence;
+            queue.remove(batch);
+        }
     }
 
     synchronized void maybeUpdateLastAckedSequence(TopicPartition topicPartition, int sequence) {
-        if (sequence > lastAckedSequence(topicPartition))
-            lastAckedSequence.put(topicPartition, sequence);
+        if (sequence > lastAckedSequence(topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER))
+            topicPartitionBookkeeper.get(topicPartition).lastAckedSequence = sequence;
     }
 
-    synchronized int lastAckedSequence(TopicPartition topicPartition) {
-        Integer currentLastAckedSequence = lastAckedSequence.get(topicPartition);
-        if (currentLastAckedSequence == null)
-            return -1;
-        return currentLastAckedSequence;
+    synchronized OptionalInt lastAckedSequence(TopicPartition topicPartition) {
+        return topicPartitionBookkeeper.lastAckedSequence(topicPartition);
     }
 
-    synchronized long lastAckedOffset(TopicPartition topicPartition) {
-        Long offset = lastAckedOffset.get(topicPartition);
-        if (offset == null)
-            return ProduceResponse.INVALID_OFFSET;
-        return offset;
+    synchronized OptionalLong lastAckedOffset(TopicPartition topicPartition) {
+        return topicPartitionBookkeeper.lastAckedOffset(topicPartition);
     }
 
     synchronized void updateLastAckedOffset(ProduceResponse.PartitionResponse response, ProducerBatch batch) {
         if (response.baseOffset == ProduceResponse.INVALID_OFFSET)
             return;
         long lastOffset = response.baseOffset + batch.recordCount - 1;
-        if (lastOffset > lastAckedOffset(batch.topicPartition)) {
-            lastAckedOffset.put(batch.topicPartition, lastOffset);
+        OptionalLong lastAckedOffset = lastAckedOffset(batch.topicPartition);
+        // It might happen that the TransactionManager has been reset while a request was reenqueued and got a valid
```

Comment:
`SenderTest.testResetOfProducerStateShouldAllowQueuedBatchesToDrain` does an example for this. I was debating a lot wether the test tests a non-existent/invalid scenario or not. The test seems valid though. Basically it sends two request, one will be rejected with an "not leader for partition" while the other one will be rejected with "out of order sequence number". The first will be reenqueued while the other one causes the transaction state to reset. Since the out of order response arrived later, the reenqueued request will be resent in the next `run()`.
Now based on what you're saying, perhaps it would be better to
- either ignore the last acked offset in this case as the response belonged to the previous producer ID which might be still valid on the broker (it hasn't timed out or been deleted). If the producer ID not valid anymore, then broker won't accept the message anyway. The only downside is that it's hard to validate if we're in this scenario.
- or to reject the message entirely saying that the producerId has changed since. This would be something like the behavior in `Sender.completeBatch` therefore may be better.

What do you think?

Reply:
I was thinking more along the lines of the first option. From a high level, when we reset the producerId, we want to start from a clean slate. Basically we want to ignore responses to requests which used the old producerId. These cases are tricky to think about. In KIP-360, I proposed that we should not reset the producerId, but should bump the epoch instead. That would make the UNKNOWN_PRODUCER_ID case unambiguous.

I think I'd suggest that we leave this case handled as it is currently so that we can merge this refactor. Then we can think about the proper handling in a separate issue. If KIP-360 is adopted, maybe this logic ends up going away anyway. Sound good?

Reply:
Sounds good to me.